pkg/controller/direct/secretmanager/secret_controller.go

```diff
@@ -0,0 +1,381 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var id *SecretManagerSecretIdentity
+	externalRef := direct.ValueOf(obj.Status.ExternalRef)
+	if externalRef == "" {
+		id = BuildID(projectID, resourceID)
+	} else {
+		id, err = asID(externalRef)
+		if err != nil {
+			return nil, err
+		}
+
+		if id.Parent.Project != projectID {
+			return nil, fmt.Errorf("SecretManagerSecret %s/%s has spec.projectRef changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Parent.Project, projectID)
+		}
+		if id.Secret != resourceID {
+			return nil, fmt.Errorf("SecretManagerSecret  %s/%s has metadata.name or spec.resourceID changed, expect %s, got %s",
+				u.GetNamespace(), u.GetName(), id.Secret, resourceID)
+		}
+	}
+
+	if err = normalizeExternal(ctx, reader, u, obj); err != nil {
+		return nil, err
+	}
+
+	// Get secretmanager GCP client
+	gcpClient, err := m.client(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &Adapter{
+		id:        id,
+		gcpClient: gcpClient,
+		desired:   obj,
+	}, nil
+}
+
+func (m *model) AdapterForURL(ctx context.Context, url string) (directbase.Adapter, error) {
+	// TODO: Support URLs
+	return nil, nil
+}
+
+type Adapter struct {
+	id        *SecretManagerSecretIdentity
+	gcpClient *gcp.Client
+	desired   *krm.SecretManagerSecret
+	actual    *secretmanagerpb.Secret
+}
+
+var _ directbase.Adapter = &Adapter{}
+
+func normalizeExternal(ctx context.Context, reader client.Reader, src client.Object, secret *krm.SecretManagerSecret) error {
+	if secret.Spec.Replication != nil {
+		if secret.Spec.Replication.LegacyAutomatic != nil {
+			if secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption != nil {
+				kmsKeyRef := secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption.KmsKeyRef
+
+				kmsKeyRef, err := refs.ResolveKMSCryptoKeyRef(ctx, reader, src, kmsKeyRef)
+				if err != nil {
+					return err
+				}
+				secret.Spec.Replication.LegacyAutomatic.CustomerManagedEncryption.KmsKeyRef = kmsKeyRef
+			}
+		}
+	}
+	if len(secret.Spec.TopicRefs) != 0 {
+		for _, topicRef := range secret.Spec.TopicRefs {
+			if topicRef.PubSubTopicRef != nil {
+				pubsubRef, err := refs.ResolvePubSubTopic(ctx, reader, src, topicRef.PubSubTopicRef)
+				if err != nil {
+					return err
+				}
+				topicRef.PubSubTopicRef.External = pubsubRef.String()
+			}
+		}
+	}
+	return nil
+
+}
+
+func (a *Adapter) Find(ctx context.Context) (bool, error) {
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("getting SecretManagerSecret", "name", a.id.FullyQualifiedName())
+
+	req := &secretmanagerpb.GetSecretRequest{Name: a.id.FullyQualifiedName()}
+	secretpb, err := a.gcpClient.GetSecret(ctx, req)
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("getting SecretManagerSecret %q: %w", a.id.FullyQualifiedName(), err)
+	}
+
+	a.actual = secretpb
+	return true, nil
+}
+
+func MergeMap(a, b map[string]string) map[string]string {
```

Comment:
Nit: this might be safer/clearer if it copied the map, rather than mutating `a` in-place